shared/src/publication/Locator.ts

```diff
@@ -0,0 +1,225 @@
+  public readonly highlight?: string;
+
+  /**
+   * Creates a [Text].
+   */
+  constructor(values: { after?: string; before?: string; highlight?: string }) {
+    this.after = values.after;
+    this.before = values.before;
+    this.highlight = values.highlight;
+  }
+
+  /**
+   * Parses a [Locations] from its RWPM JSON representation.
+   */
+  public static deserialize(json: any): LocatorText | undefined {
+    if (!json) return;
+    return new LocatorText({
+      after: json.after,
+      before: json.before,
+      highlight: json.highlight,
+    });
+  }
+
+  /**
+   * Serializes a [Locations] to its RWPM JSON representation.
+   */
+  public serialize(): any {
+    const json: any = {};
+    if (this.after !== undefined) json.after = this.after;
+    if (this.before !== undefined) json.before = this.before;
+    if (this.highlight !== undefined) json.highlight = this.highlight;
+    return json;
+  }
+}
+
+/**
+ * Provides a precise location in a publication in a format that can be stored and shared.
+ *
+ * There are many different use cases for locators:
+ *  - getting back to the last position in a publication
+ *  - bookmarks
+ *  - highlights & annotations
+ *  - search results
+ *  - human-readable (and shareable) reference in a publication
+ *
+ * https://github.com/readium/architecture/tree/master/locators
+ */
+export class Locator {
+  /** The URI of the resource that the Locator Object points to. */
+  public readonly href: string;
+
+  /** The media type of the resource that the Locator Object points to. */
+  public readonly type: string;
+
+  /** The title of the chapter or section which is more relevant in the context of this locator. */
+  public readonly title?: string;
+
+  /** One or more alternative expressions of the location. */
+  public readonly locations: LocatorLocations;
+
+  /** Textual context of the locator. */
+  public readonly text?: LocatorText;
+
+  /**
+   * Creates a [Locator].
+   */
+  constructor(values: {
+    href: string;
+    type: string;
+    title?: string;
+    locations?: LocatorLocations;
+    text?: LocatorText;
+  }) {
+    this.href = values.href;
+    this.type = values.type;
+    this.title = values.title;
+    this.locations = values.locations
+      ? values.locations
+      : new LocatorLocations({});
+    this.text = values.text;
+  }
+
+  /**
+   * Parses a [Link] from its RWPM JSON representation.
+   */
+  public static deserialize(json: any): Locator | undefined {
+    if (!(json && json.href && json.type)) return;
+    return new Locator({
+      href: json.href,
+      type: json.type,
+      title: json.title,
+      locations: LocatorLocations.deserialize(json.locations),
+      text: LocatorText.deserialize(json.text),
+    });
+  }
+
+  /**
+   * Serializes a [Link] to its RWPM JSON representation.
+   */
+  public serialize(): any {
+    const json: any = { href: this.href, type: this.type };
+    if (this.title !== undefined) json.title = this.title;
```

Comment:
As a pattern in your serializer functions, it is safe to assign `undefined` to variables that are already `undefined`, so you can remove many of your `if` guards.

```suggestion
    json.title = this.title;
```